### PR TITLE
Add support for libusb device whitelisting

### LINKED ISSUE #6005: [hidapi] Whitelisting for libusb-only devices
We've had support for Nintendo's official GameCube adapter for a while...

https://github.com/libsdl-org/SDL/blob/main/src/joystick/hidapi/SDL_hidapi_gamecube.c

... but the big problem that's always kept it locked away is that it _requires_ libusb. This is because while it _looks_ like a HID, it doesn't _act_ like one: You have to manually initialize it with a magic number, then it suddenly starts acting like a real input device. This means the normal paths are out of the question (with the exception of writing support into the kernel, which is way out of my league and I don't know anyone else that wants to do it), and since libusb is [too aggressive to enable by default](https://github.com/libsdl-org/SDL/issues/5722#issuecomment-1139748983) most developers ship without it.

But from what I recall, it only takes exclusive control once you open the device, not during enumeration (unless you go out of your way to enable a flag that detaches drivers during enumeration). On top of that, the way we enumerate in SDL is by [getting the libusb devices first, then the platform implementation next](https://github.com/libsdl-org/SDL/blob/main/src/hidapi/SDL_hidapi.c#L1172). So I wonder if, for platforms with non-libusb backends, we could make a whitelist for devices that require libusb and then ignore the rest of what the backend finds, with the expectation that the platform API will pick those up the Proper Way(TM).

Odds are this would just be for WUP-028, but I wonder if there are other devices out there like this? In any case, this seems like the best way to achieve [the ultimate goal for SteamOS controller support](https://nintendo.fandom.com/wiki/DK_Bongo) provided the idea actually works.

CC: @slouken

### PR DESCRIPTION
**Pull request by flibitijibibo**
When combining both libusb and an OS backend for SDL_hidapi, we want to ensure that libusb is only used when absolutely necessary. At the same time, we don't want to interfere with platforms that _only_ have libusb. So, implement a whitelist but only use it when another backend exists.

With this in place, it should now be safe to enable libusb support by default.

Affects SDL_hidapi (Sam), configure.ac (Ozkan), and CMake (Anon).

Fixes #6005

### CODE REVIEW

**Comment by sezero** [comment]
This doesn't bring back https://github.com/libsdl-org/SDL/issues/5722 does it?

**Comment by flibitijibibo** [comment]
No, it should only affect devices that aren't visible to the Linux hid.c backend. There's logic in the libusb hid.c that takes over devices during enumeration but it's disabled (with good reason).

File: src/hidapi/SDL_hidapi.c [hunk 1]
```diff
@@ -1197,6 +1215,23 @@ struct SDL_hid_device_info *SDL_hid_enumerate(unsigned short vendor_id, unsigned
...
         for (usb_dev = usb_devs; usb_dev; usb_dev = usb_dev->next) {
+#if USE_LIBUSB_WHITELIST
+            int i;
+            for (i = 0; i < SDL_arraysize(SDL_libusb_whitelist); i += 1) {
+                if (usb_dev->vendor_id == SDL_libusb_whitelist[i].vendor &&
+                    usb_dev->product_id == SDL_libusb_whitelist[i].product) {
+                    break;
+                }
+            }
+            if (i == SDL_arraysize(SDL_libusb_whitelist)) {
```
Comment (slouken) [inline src/hidapi/SDL_hidapi.c:1226]: Suggest using `SDL_bool in_whitelist;` or something similar, for clarity.

**Comment by slouken** [comment]
I believe this will break Xbox controllers on Steam Link, where we use linux/hid.c, but xpad is largely unused and we rely on libusb for access to those devices.

I'm putting this in the next milestone so we can smoke out issues like these early in the milestone.

**Comment by flibitijibibo** [comment]
Pushed the SDL_bool change.

For Link, it should be possible to make the list user-modifiable, not sure how that might be done though.

**Comment by flibitijibibo** [comment]
Moved the check a little closer to the whitelist declaration, to make it easier to keep the whitelist work in one place (in case we do end up adding variables or something).

**Comment by slouken** [comment]
> For Link, it should be possible to make the list user-modifiable, not sure how that might be done though.

For systems where exclusive access to the device isn't an issue, like Steam Link, we probably want to turn this off. Maybe we can make it a configure option?

**Comment by slouken** [comment]
It's almost like we want 3 passes:
1. Devices where libusb is preferred (e.g. GameCube adapter)
2. Devices where platform support is preferred (e.g. almost everything else)
3. Devices which platform support doesn't pick up (e.g. Xbox controllers)

**Comment by flibitijibibo** [comment]
If it's possible to prioritize, switching the order to have libusb as the last pass would definitely be enough, since the adapter doesn't show up as a regular device anyhow.

**Comment by slouken** [comment]
It looks like the backend priority was flipped to libusb first in this commit:
https://github.com/libsdl-org/SDL/commit/69a8c8468c211d66ad547a352c4762470298aa36

**Comment by slouken** [comment]
libusb enforces exclusive access to the device, which isn't generally useful on desktop systems. I don't mind a define for custom systems that enables this, but I don't think we should do it by default.

Feel free to submit another PR that implements an option for this, but I'll go ahead and close this one.

**Comment by flibitijibibo** [comment]
Just to be sure: Does it enforce exclusive access during enumeration time? The main goal of this patch was to enumerate first and avoid any device that wasn't in the list, so that libusb doesn't take it over - if anything this should make it take devices _less_ often, but that will only be true if enumeration doesn't grab the device to get the product information.

**Comment by slouken** [comment]
No, enumeration is fine, it's only if you open it. I'm going to reopen this, I think I misunderstood the intent.

**Comment by slouken** [comment]
I'm going to change it to merge lists based on device priority, but I haven't gotten time to do that yet.

**Comment by flibitijibibo** [comment]
I'll rebase this today!

File: include/build_config/SDL_build_config.h.cmake [hunk 2]
```diff
@@ -330,6 +330,7 @@
 #cmakedefine SDL_HAPTIC_DINPUT @SDL_HAPTIC_DINPUT@
 #cmakedefine SDL_HAPTIC_XINPUT @SDL_HAPTIC_XINPUT@
 #cmakedefine SDL_HAPTIC_ANDROID @SDL_HAPTIC_ANDROID@
+#cmakedefine SDL_LIBUSB_WHITELIST @SDL_LIBUSB_WHITELIST@
```
Comment (slouken) [inline include/build_config/SDL_build_config.h.cmake:333]: Where is this define used?

**Comment by Lokathor** [comment]
Just to be clear:
* would this make the pre-built SDL for windows support things like the GameCube controller automatically?
* would an SDL built from source also need to get a copy of libusb for windows during the build for that to be enabled, or is the necessary libusb stuff dynamically linked or whatever?

**Comment by flibitijibibo** [comment]
Updated to "fix" the CMake changes, but my solution is really clumsy - feel free to redo it if it's not too hard to fix! (This may affect the premade project files as well, not sure if that's in CI...)

**Comment by slouken** [comment]
@madebr, can you review the CMake changes?

**Comment by slouken** [comment]
> Just to be clear:
> 
> * would this make the pre-built SDL for windows support things like the GameCube controller automatically?
> * would an SDL built from source also need to get a copy of libusb for windows during the build for that to be enabled, or is the necessary libusb stuff dynamically linked or whatever?

I don't think we're planning on adding libusb to the Windows build. That may change by the time SDL 3.0 ships though.

**Comment by Lokathor** [comment]
Ah. It was suggested in https://github.com/libsdl-org/SDL/issues/7481 that this PR would be the most likely path to using gamecube controllers on Windows. Unfortunate, but I guess that can be resolved in the future.

**Comment by flibitijibibo** [comment]
Took a different approach that mostly avoids CMake altogether - instead of determining the use as a build config, I've changed it to a hint, which is enabled by default _only_ when there are other hidapi backends available. This seemed like a better way to allow enabling/disabling this for a particular application without having to rebuild SDL or redirect dynapi to a custom file.

**Comment by madebr** [comment]
Assuming https://github.com/libsdl-org/SDL/pull/6014#issuecomment-1207251777 holds, the cmake changes are fine: libusb will be enabled for most platforms.

**Comment by slouken** [comment]
Looks good, thanks!

**Comment by slouken** [comment]
This did indeed break Steam Link hardware:
https://steamcommunity.com/app/353380/discussions/0/5992658048661696817/

The problem is that opening a device with libusb unbinds it from the HID subsystem, which is what SDL uses to detect and open Steam Controllers on Steam Link hardware.

What we really need to do is enumerate the devices using all the back ends and then add them to the enumeration list based on driver priority, which is controlled by the whitelisting process.

**Comment by slouken** [comment]
Fixed in 5774c9638

**Pull request merged**